src/jit.c

```diff
@@ -0,0 +1,1578 @@
+/*
+ * rv32emu is freely redistributable under the MIT License. See the file
+ * "LICENSE" for information on usage and redistribution of this file.
+ */
+
+/* This JIT implementation has undergone extensive modifications, heavily
+ * relying on the ubpf_jit_[x86_64|arm64].[ch] from ubpf. The original
+ * ubpf_jit_[x86_64|arm64].[ch] file served as the foundation and source of
+ * inspiration for adapting and tailoring it specifically for this JIT
+ * implementation. Therefore, credit and sincere thanks are extended to ubpf for
+ * their invaluable work.
+ *
+ * Reference:
+ *   https://github.com/iovisor/ubpf/blob/main/vm/ubpf_jit_x86_64.c
+ *   https://github.com/iovisor/ubpf/blob/main/vm/ubpf_jit_arm64.c
+ */
+
+#if !defined(__x86_64__) && !defined(__aarch64__)
+#error "This implementation is dedicated to x64 and arm64."
+#endif
+
+#include <assert.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#if HAVE_MMAP
+#include <sys/mman.h>
+#include <unistd.h>
+#endif
+#if defined(__APPLE__) && defined(__aarch64__)
+#include <libkern/OSCacheControl.h>
+#include <pthread.h>
+#endif
+
+#include "cache.h"
+#include "decode.h"
+#include "io.h"
+#include "jit.h"
+#include "riscv.h"
+#include "utils.h"
+
+#define JIT_CLS_MASK 0x07
+#define JIT_ALU_OP_MASK 0xf0
+#define JIT_CLS_ALU 0x04
+#define JIT_CLS_ALU64 0x07
+#define JIT_SRC_IMM 0x00
+#define JIT_SRC_REG 0x08
+#define JIT_OP_MUL_IMM (JIT_CLS_ALU | JIT_SRC_IMM | 0x20)
+#define JIT_OP_MUL_REG (JIT_CLS_ALU | JIT_SRC_REG | 0x20)
+#define JIT_OP_DIV_IMM (JIT_CLS_ALU | JIT_SRC_IMM | 0x30)
+#define JIT_OP_DIV_REG (JIT_CLS_ALU | JIT_SRC_REG | 0x30)
+#define JIT_OP_MOD_IMM (JIT_CLS_ALU | JIT_SRC_IMM | 0x90)
+#define JIT_OP_MOD_REG (JIT_CLS_ALU | JIT_SRC_REG | 0x90)
+
+#define STACK_SIZE 512
+#define MAX_INSNS 1024
+#if defined(__x86_64__)
+#define JUMP_LOC jump_loc + 2
+/* Special values for target_pc in struct jump */
+#define TARGET_PC_EXIT -1U
+#define TARGET_PC_RETPOLINE -3U
+enum X64_REG {
+    RAX,
+    RCX,
+    RDX,
+    RBX,
+    RSP,
+    RBP,
+    RIP = 5,
+    RSI,
+    RDI,
+    R8,
+    R9,
+    R10,
+    R11,
+    R12,
+    R13,
+    R14,
+    R15,
+};
+
+#elif defined(__aarch64__)
+#define JUMP_LOC jump_loc
+/* Special values for target_pc in struct jump */
+#define TARGET_PC_EXIT ~UINT32_C(0)
+#define TARGET_PC_ENTER (~UINT32_C(0) & 0x0101)
+/* This is guaranteed to be an illegal A64 instruction. */
+#define BAD_OPCODE ~UINT32_C(0)
+
+enum A64_REG {
+    R0,
+    R1,
+    R2,
+    R3,
+    R4,
+    R5,
+    R6,
+    R7,
+    R8,
+    R9,
+    R10,
+    R11,
+    R12,
+    R13,
+    R14,
+    R15,
+    R16,
+    R17,
+    R18,
+    R19,
+    R20,
+    R21,
+    R22,
+    R23,
+    R24,
+    R25,
+    R26,
+    R27,
+    R28,
+    R29,
+    R30,
+    SP,
+    RZ = 31
+};
+
+enum Opcode {
```

Comment:
Always use lowercase for the enum name unless you have strong reasons as an exception.